adapters/unicorn/unicorn.go

```diff
@@ -0,0 +1,285 @@
+)
+
+// UnicornAdapter describes a UNICORN prebid server adapter.
+type UnicornAdapter struct {
+	endpoint string
+}
+
+// unicornImpExt is imp ext for UNICORN
+type unicornImpExt struct {
+	Context *unicornImpExtContext     `json:"context,omitempty"`
+	Bidder  openrtb_ext.ExtImpUnicorn `json:"bidder"`
+}
+
+type unicornImpExtContext struct {
+	Data interface{} `json:"data,omitempty"`
+}
+
+// unicornSourceExt is source ext for UNICORN
+type unicornSourceExt struct {
+	Stype  string `json:"stype"`
+	Bidder string `json"bidder"`
+}
+
+// unicornExt is ext for UNICORN
+type unicornExt struct {
+	Prebid    *openrtb_ext.ExtImpPrebid `json:"prebid,omitempty"`
+	AccountID int64                     `json:"accountId,omitempty"`
+}
+
+// Builder builds a new instance of the Foo adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	bidder := &UnicornAdapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+// MakeRequests makes the HTTP requests which should be made to fetch bids.
+func (a *UnicornAdapter) MakeRequests(request *openrtb.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var extRegs openrtb_ext.ExtRegs
+	if request.Regs != nil {
+		if request.Regs.COPPA == 1 {
+			return nil, []error{&errortypes.BadInput{
+				Message: "COPPA is not supported",
+			}}
+		}
+		if err := json.Unmarshal(request.Regs.Ext, &extRegs); err == nil {
+			if extRegs.GDPR != nil && (*extRegs.GDPR == 1) {
+				return nil, []error{&errortypes.BadInput{
+					Message: "GDPR is not supported",
+				}}
+			}
+			if extRegs.USPrivacy != "" {
+				return nil, []error{&errortypes.BadInput{
+					Message: "CCPA is not supported",
+				}}
+			}
+		}
+	}
+
+	imp, err := setImps(request, requestInfo)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	request.Imp = imp
+
+	request.Source.Ext, err = setSourceExt()
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	request.Ext, err = setExt(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    requestJSON,
+		Headers: getHeaders(request),
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func getHeaders(request *openrtb.BidRequest) http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IPv6) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IPv6)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+	}
```

Comment:
Please add test coverage for the existence of UA, IPv6, and IP.

Reply:
Fixed.